### PR TITLE
Add koa-tree-router, fix other routers

### PR DESCRIPTION
**Pull request by alecmev**
Added [`koa-tree-router`](https://github.com/steambap/koa-tree-router), since I was curious how it stacks up against `koa-router` and `find-my-way`. The results on my i7-8550U CPU + 1866MHz RAM + Linux 4.15 + Node 9.6 are at the end of this post. Please, let me know if you're okay with me replacing the ones in the `README.md` with these.

While doing that, I have noticed a few issues.

The biggest one is the fact that the only apples-to-apples benchmarks are `call`, `find-my-way` and `routr`. The rest are doing extra work (handling and such) and `koa-router` is outright invalid. I have attempted fixing what I could, and added a warning where it isn't possible (can't decouple lookup from handling in `express` and `router`). Could do the opposite (i.e. include handling in all tests), but that isn't framework-agnostic, and clearly isn't your intent, since the router you're the creator of - `find-my-way` - isn't doing that in its benchmark. On the other hand, how does one compare it to `express` in a fair way then?

Another important issue is the fact that this suite isn't very consistent/accurate, measurement-wise. I see you're [using `benchmark` in `find-my-way`](https://github.com/delvedor/find-my-way/blob/ba04e309b28b64f18fe6a535b6dad5e1ee1f89e8/bench.js), why not use it here? And on a related note, would be nice to have a right-aligned sorted list of "all together" ops/sec from all routers, and possibly an accompanying bar graph, and some sort of overall "editorial" ranking.

Lastly, the benchmarks are way too WET. Have you considered taking a more declarative approach? Have a preparation step, transforming a common `routes` list (probably express-style, since half the routers support it as-is) into a router-specific format, and then an execution step, invoking a given lookup function with the method and path passed, again, in a router-specific manner. Otherwise I can see subtle differences starting to creep in between the benchmarks, as more of them are being added/updated, skewing the results.

Without addressing the above, IMO, this suite shouldn't be used as `find-my-way`'s advertisement.

```
=======================
 find-my-way benchmark
=======================
short static: 13,810,332 ops/sec
static with same radix: 6,273,479 ops/sec
dynamic route: 1,682,494 ops/sec
mixed static dynamic: 2,006,188 ops/sec
long static: 8,335,677 ops/sec
wildcard: 2,693,026 ops/sec
all together: 498,427 ops/sec

================
 call benchmark
================
short static: 3,700,535 ops/sec
static with same radix: 3,598,872 ops/sec
dynamic route: 792,128 ops/sec
mixed static dynamic: 862,072 ops/sec
long static: 3,763,495 ops/sec
wildcard: 1,292,482 ops/sec
all together: 232,753 ops/sec

================================================
 express benchmark (WARNING: includes handling)
================================================
short static: 1,217,991 ops/sec
static with same radix: 1,134,697 ops/sec
dynamic route: 723,241 ops/sec
mixed static dynamic: 628,484 ops/sec
long static: 636,273 ops/sec
wildcard: 500,932 ops/sec
all together: 105,345 ops/sec

======================
 koa-router benchmark
======================
short static: 1,137,411 ops/sec
static with same radix: 1,172,908 ops/sec
dynamic route: 1,115,376 ops/sec
mixed static dynamic: 1,101,533 ops/sec
long static: 1,179,059 ops/sec
wildcard: 1,117,365 ops/sec
all together: 177,765 ops/sec

===========================
 koa-tree-router benchmark
===========================
short static: 12,929,070 ops/sec
static with same radix: 6,047,950 ops/sec
dynamic route: 2,984,655 ops/sec
mixed static dynamic: 3,787,085 ops/sec
long static: 7,840,458 ops/sec
wildcard: 4,106,017 ops/sec
all together: 781,418 ops/sec

===============================================
 router benchmark (WARNING: includes handling)
===============================================
short static: 1,176,093 ops/sec
static with same radix: 1,045,083 ops/sec
dynamic route: 685,231 ops/sec
mixed static dynamic: 589,179 ops/sec
long static: 623,477 ops/sec
wildcard: 467,692 ops/sec
all together: 94,245 ops/sec

=================
 routr benchmark
=================
short static: 3,951,940 ops/sec
static with same radix: 1,975,343 ops/sec
dynamic route: 853,571 ops/sec
mixed static dynamic: 537,769 ops/sec
long static: 409,445 ops/sec
wildcard: 323,400 ops/sec
all together: 101,739 ops/sec

=========================
 server-router benchmark
=========================
short static: 2,837,363 ops/sec
static with same radix: 2,902,904 ops/sec
dynamic route: 1,677,476 ops/sec
mixed static dynamic: 1,807,669 ops/sec
long static: 1,849,464 ops/sec
wildcard: 1,454,016 ops/sec
all together: 284,792 ops/sec
```

### CODE REVIEW

**Comment by delvedor** [comment]
Hi! Thank you for contributing and checking out the repo!
Sorry for the delay, but this repo is not my top priority and it has been a rough month :P

> The biggest one is the fact that the only apples-to-apples benchmarks are call, find-my-way and routr. The rest are doing extra work (handling and such) and koa-router is outright invalid. I have attempted fixing what I could, and added a warning where it isn't possible (can't decouple lookup from handling in express and router). Could do the opposite (i.e. include handling in all tests), but that isn't framework-agnostic, and clearly isn't your intent, since the router you're the creator of - find-my-way - isn't doing that in its benchmark. On the other hand, how does one compare it to express in a fair way then?

You are right, thank you for fixing the others, but in the case of express there are two ways, try to extract the router or benchmark the `path-to-regexp` library which is doing the hard work.

> Another important issue is the fact that this suite isn't very consistent/accurate, measurement-wise. I see you're using benchmark in find-my-way, why not use it here?

The benchmark code used in `find-my-way` was written before this code, and I'll update it someday. Currently is just used to verify that there are not regressions between new prs and master.

>  And on a related note, would be nice to have a right-aligned sorted list of "all together" ops/sec from all routers, and possibly an accompanying bar graph, and some sort of overall "editorial" ranking.

Because I wrote this in my spare time, but contributions are welcome!

> Lastly, the benchmarks are way too WET. Have you considered taking a more declarative approach? Have a preparation step, transforming a common routes list (probably express-style, since half the routers support it as-is) into a router-specific format, and then an execution step, invoking a given lookup function with the method and path passed, again, in a router-specific manner. Otherwise I can see subtle differences starting to creep in between the benchmarks, as more of them are being added/updated, skewing the results.

You are right, but see the above response :P
In any case I wanted a system with the minimum overhead when generating the benchmarks, since the results could be affected by many things.

> Without addressing the above, IMO, this suite shouldn't be used as find-my-way's advertisement.

This is not an *advertisement* I created this repo just to see the performances of `find-my-way` compared to others **commonly used** routers, and see if we can do better.

File: benchmarks/server-router.js [hunk 1]
```diff
@@ -30,47 +30,47 @@ routes.forEach(route => {
 
 time = now()
 for (i = 0; i < operations; i++) {
-  router.match({ method: 'GET', url: '/user' })
+  router._router.match('GET/user')
```
Comment (delvedor) [inline benchmarks/server-router.js:33]: Are you sure this is correct?
`server-router` internally uses `wayfarer` which is not an http router, that's why is not present in the benchmarks.

**Comment by alecmev** [comment]
Hi, thanks for dedicating time to this!

> in the case of express there are two ways, try to extract the router or benchmark the path-to-regexp library which is doing the hard work

> server-router internally uses wayfarer which is not an http router, that's why is not present in the benchmarks

This begs the question of what is an HTTP router, and what isn't? Where's the line? Wayfarer kind of fits the bill (that is "method handling should be included"), `server-router` is merely a thin convenience wrapper around it. Maybe handling should be tested too then, to highlight any inefficiencies in the way the handlers are invoked? With mock `req` and `res`.

>> Another important issue is the fact that this suite isn't very consistent/accurate, measurement-wise. I see you're using benchmark in find-my-way, why not use it here?
>
> The benchmark code used in find-my-way was written before this code, and I'll update it someday. Currently is just used to verify that there are not regressions between new prs and master.

To be clear, I was pointing at the usage of [Benchmark.js](https://github.com/bestiejs/benchmark.js/).

> This is not an advertisement I created this repo just to see the performances of find-my-way compared to others commonly used routers, and see if we can do better.

I understand that! And thanks for making it, it helped me make a decision. However, I'm referring to [this sentence](https://github.com/delvedor/find-my-way/blame/ba04e309b28b64f18fe6a535b6dad5e1ee1f89e8/README.md#L7) in `find-my-way`'s readme, inviting people to compare its performance to other routers. Some might take the results at face value and conclude that it's ~2+ times faster than the competition (and whopping ~5 times faster than `express`, as seen on my machine), which most likely isn't representative of the real-world performance differences.

> Because I wrote this in my spare time, but contributions are welcome!

> You are right, but see the above response :P

It's all good! Just wishes/suggestions. I'd probably enjoy the refactor, but can't afford spending more spare time beyond this PR either, sorry.

File: benchmarks/server-router.js (hunk 1 above)
Comment (alecmev) [reply benchmarks/server-router.js:33]: As mentioned in the PR comment, I actually think `wayfarer` qualifies as an HTTP router. It just needs a little bit of special care with HTTP verbs, which `server-router` [abstracts away in the handling mode](https://github.com/yoshuawuyts/server-router/blob/811aaaaaa93a35c039a21b637fc02ef6d6fcb35a/index.js#L37-L41).

**Comment by delvedor** [comment]
> This begs the question of what is an HTTP router, and what isn't? Where's the line? Wayfarer kind of fits the bill (that is "method handling should be included"), server-router is merely a thin convenience wrapper around it.

I think is pretty obvious. What do you expect from an **http** router (not a basic router)?
It should be able to handle the request method, static and dynamic routes and also wildcards.
Wayfarer is just handling the *path* part, but handle also the method comes with a cost.

>Maybe handling should be tested too then, to highlight any inefficiencies in the way the handlers are invoked? With mock req and res.

This could be an idea, even better check both things!
One benchmark just for the "find" and the other for the handling.

> To be clear, I was pointing at the usage of Benchmark.js.

I know, still same answer.

 > I understand that! And thanks for making it, it helped me make a decision. However, I'm referring to this sentence in find-my-way's readme, inviting people to compare its performance to other routers. Some might take the results at face value and conclude that it's ~2+ times faster than the competition (and whopping ~5 times faster than express, as seen on my machine), which most likely isn't representative of the real-world performance differences.

Express will be slow even if we can figure out a way to remove the handling. It uses regular expressions, which are very slow.

**Comment by alecmev** [comment]
> Wayfarer is just handling the path part, but handle also the method comes with a cost.

Wayfarer does do handling:

```js
var router = wayfarer();
router.on('GET/:user', ({ user }, req, res) => console.log(`user is ${user}`, req, res));
router('GET/someUser', req, res);
```

Just in a non-conventional way, which is where `server-router` comes in.

> > To be clear, I was pointing at the usage of Benchmark.js.
>
> I know, still same answer.

But if you wrote this suite *after* `find-my-way`'s benchmark, then why didn't you use Benchmark.js? Or do you dislike it or find it unfitting?

> Express will be slow even if we can figure out a way to remove the handling. It uses regular expressions, which are very slow.

I don't doubt that, it just won't be a factor of 5, at least with handling.

**Comment by delvedor** [comment]
> Just in a non-conventional way, which is where server-router comes in.

Maybe this could work here, but I'm not sure it will cover all the possible issues.
When I will have time I'll update the benchmarks and test the handlers separately.
In any case the handler should not add too much overhead, at lest no so much to justify a significant drop in performances.

> But if you wrote this suite after find-my-way's benchmark, then why didn't you use Benchmark.js? Or do you dislike it or find it unfitting?

I've dropped Benchmarkjs here because I wanted to have total control on the benchmark runner, in this way in future iterations of this suite I will not have problems with the api of benchmarkjs.

> I don't doubt that, it just won't be a factor of 5, at least with handling.

Trust me, is **really** slow.

**Comment by alecmev** [comment]
> I've dropped Benchmarkjs here because I wanted to have total control on the benchmark runner, in this way in future iterations of this suite I will not have problems with the api of benchmarkjs.

But it does a lot more under the hood than just timing fixed-iteration loops, hence why I mentioned consistency/accuracy. Take a look at [the article](https://mathiasbynens.be/notes/javascript-benchmarking) they link in their readme.

> Trust me, is really slow.

I know :smile: That's why I use Koa. But, just for the record, merely replacing `find` with `lookup` makes the `find-my-way` bench 35% slower, putting it at 3.2x speed of `express`, instead of ~5x.

**Comment by delvedor** [comment]
> But it does a lot more under the hood than just timing fixed-iteration loops, hence why I mentioned consistency/accuracy. Take a look at the article they link in their readme.

I'll read it, thanks.

**Pull request merged**